pkg/reconciler/workspaceindex/orchestrator.go

```diff
@@ -0,0 +1,506 @@
+		runtime.HandleError(fmt.Errorf("%q controller failed to sync %q, err: %w", controllerName, key, err))
+		o.orgQueue.AddRateLimited(key)
+		return true
+	}
+	o.orgQueue.Forget(key)
+	return true
+}
+
+func (o *Orchestrator) processOrg(ctx context.Context, key string) error {
+	obj, err := o.workspaceLister.Get(key) // TODO: clients need a way to scope down the lister per-cluster
+	if err != nil {
+		if errors.IsNotFound(err) {
+			return nil // object deleted before we handled it
+		}
+		return err
+	}
+	obj = obj.DeepCopy()
+
+	return o.handleOrg(ctx, obj)
+}
+
+// handleOrg ensures that an informer is running for Workspaces in every Organization, using the
+// latest credentials available for the Shard on which the Organization is held
+func (o *Orchestrator) handleOrg(ctx context.Context, workspace *tenancyv1alpha1.Workspace) error {
+	if !conditions.IsTrue(workspace, tenancyv1alpha1.WorkspaceScheduled) ||
+		!conditions.IsTrue(workspace, tenancyv1alpha1.WorkspaceURLValid) {
+		klog.Infof("org workspace %s/%s not ready, skipping...", workspace.ClusterName, workspace.Name)
+		return nil // not ready yet
+	}
+	shardKey, err := cache.MetaNamespaceKeyFunc(&metav1.ObjectMeta{
+		ClusterName: workspace.ObjectMeta.ClusterName,
+		Name:        workspace.Status.Location.Current,
+	})
+	if err != nil {
+		klog.Errorf("failed to get key for workspace shard: %v", err)
+		return nil
+	}
+	workspaceShard, err := o.workspaceShardLister.Get(shardKey)
+	if err != nil {
+		klog.Errorf("failed to get workspace shard %q: %v", shardKey, err)
+		return nil
+	}
+
+	var previous string
+	o.informerLock.Lock()
+	defer o.informerLock.Unlock()
+	if i, ok := o.shardInformers[workspace.Status.Location.Current]; ok {
+		previous = i.credentialsHash
+	}
+
+	if previous == workspaceShard.Status.CredentialsHash {
+		// nothing to be done
+		return nil
+	}
+
+	secret, err := o.kubeClient.Cluster(workspaceShard.ClusterName).CoreV1().Secrets(workspaceShard.Spec.Credentials.Namespace).Get(ctx, workspaceShard.Spec.Credentials.Name, metav1.GetOptions{})
+	if err != nil {
+		klog.Errorf("failed to get workspace shard credentials: %v", shardKey, err)
+		return err // retry since this could flake in an otherwise OK situation
+	}
+
+	data, ok := secret.Data[tenancyv1alpha1.WorkspaceShardCredentialsKey]
+	if !ok {
+		klog.Error("workspace shard credentials missing data")
+		return nil
+	}
+
+	cfg, err := clientcmd.RESTConfigFromKubeConfig(data)
+	if err != nil {
+		klog.Errorf("workspace shard credentials invalid: %v", err)
+		return nil
+	}
+
+	if previous != "" {
+		klog.Info("oops")
+		// TODO: figure out current resourceVersion from the previous lister and start new one there
+	}
+
+	const clusterAll = "*"
+	kcpClient, err := kcpclient.NewClusterForConfig(cfg)
+	if err != nil {
+		klog.Errorf("error creating client with workspace shard credentials: %v", err)
+		return nil
+	}
+
+	octx, cancel := context.WithCancel(ctx)
+	crossClusterKcpClient := kcpClient.Cluster(clusterAll)
```

Comment:
clusterAll is surprising here. I would have expected to watch just the new org workspace.

Reply:
... whose lcluster we know.